sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -797,6 +854,61 @@ public Mono<List<Address>> getMasterAddressesViaGatewayAsync(
         });
     }
 
+    /***
+     *  merge the new addresses get back from gateway with the cached addresses.
+     *  If the address is being returned from gateway again, then keep using the cached addressInformation object
+     *  If it is a new address being returned, then use the new addressInformation object.
+     *
+     * @param newAddresses the latest addresses being returned from gateway.
+     * @param cachedAddresses the cached addresses.
+     *
+     * @return the merged addresses.
+     */
+    private AddressInformation[] mergeAddresses(AddressInformation[] newAddresses, AddressInformation[] cachedAddresses) {
+        checkNotNull(newAddresses, "Argument 'newAddresses' should not be null");
+
+        if (cachedAddresses == null) {
+            return newAddresses;
+        }
+
+        List<AddressInformation> mergedAddresses = new ArrayList<>();
+        Map<Uri, AddressInformation> cachedAddressMap =
+                Arrays
+                    .stream(cachedAddresses)
+                    .collect(Collectors.toMap(address -> address.getPhysicalUri(), address -> address));
+
+        for (AddressInformation addressInformation : newAddresses) {
+            if (cachedAddressMap.containsKey(addressInformation.getPhysicalUri())) {
+                mergedAddresses.add(cachedAddressMap.get(addressInformation.getPhysicalUri()));
+            } else {
+                mergedAddresses.add(addressInformation);
+            }
+        }
+
+        return mergedAddresses.toArray(new AddressInformation[mergedAddresses.size()]);
```

Comment:
Instead of creating a new array, I think we can use the existing map to keep track of new addresses. 

```
Map<Uri, AddressInformation> cachedAddressMap =
                Arrays
                    .stream(cachedAddresses)
                    .collect(Collectors.toMap(address -> address.getPhysicalUri(), address -> address));

        for (AddressInformation addressInformation : newAddresses) {
            if (!cachedAddressMap.containsKey(addressInformation.getPhysicalUri())) {
                cachedAddressMap.put(addressInformation.getPhysicalUri(), addressInformation);
            }
        }

        return cachedAddressMap.values().toArray(new AddressInformation[0]);
```

Reply:
we also want to remove the addresses not returned in the new addresses